File: kotlin/logbrew-kotlin/src/main/kotlin/co/logbrew/sdk/LogBrewClient.kt
package co.logbrew.sdk

class LogBrewClient private constructor(
    private val apiKey: String,
    sdkName: String,
    sdkVersion: String,
    private val maxRetries: Int,
    private val context: TelemetryContext?,
) {
    private val sdk =
        OrderedJsonObject()
            .add("name", sdkName)
            .add("language", "kotlin")
            .add("version", sdkVersion)
    private val stateLock = Any()
    private val events = mutableListOf<Event>()
    private val breadcrumbStore = IssueBreadcrumbStore()
    private var closed = false

    fun pendingEvents(): Int = synchronized(stateLock) { events.size }

    fun previewJson(): String = synchronized(stateLock) { previewJsonLocked() }

    private fun previewJsonLocked(): String =
        JsonWriter.write(
            OrderedJsonObject()
                .add("sdk", sdk)
                .add("events", events.map { it.toJsonObject() }),
        )

    fun release(
        id: String,
        timestamp: String,
        attributes: ReleaseAttributes,
    ) {
        pushEvent("release", id, timestamp, attributes.copy(context = resolvedContext(attributes.context)).toJsonObject())
    }

    fun environment(
        id: String,
        timestamp: String,
        attributes: EnvironmentAttributes,
    ) {
        pushEvent("environment", id, timestamp, attributes.copy(context = resolvedContext(attributes.context)).toJsonObject())
    }

    fun issue(
        id: String,
        timestamp: String,
        attributes: IssueAttributes,
    ) {
        val snapshot = synchronized(stateLock) { breadcrumbStore.snapshot() }
        val resolvedContext = resolvedContext(attributes.context)
        val resolved =
            attributes
                .withBreadcrumbSnapshot(snapshot)
                .withResolvedTraceMetadata(resolvedContext?.trace)
                .copy(context = resolvedContext)
        pushEvent("issue", id, timestamp, resolved.toJsonObject())
    }

    fun log(
        id: String,
        timestamp: String,
        attributes: LogAttributes,
    ) {
        val resolvedContext = resolvedContext(attributes.context)
        val resolved =
            attributes
                .withResolvedTraceMetadata(resolvedContext?.trace)
                .copy(context = resolvedContext)
        pushEvent("log", id, timestamp, resolved.toJsonObject())
    }

    fun span(
        id: String,
        timestamp: String,
        attributes: SpanAttributes,
    ) {
        val resolvedContext = resolvedSpanContext(attributes.context, attributes.telemetryTraceContext())
        pushEvent("span", id, timestamp, attributes.copy(context = resolvedContext).toJsonObject())
    }

    fun metric(
        id: String,
        timestamp: String,
        attributes: MetricAttributes,
    ) {
        val resolvedContext = resolvedContext(attributes.context)
        val resolved =
            attributes
                .withResolvedTraceMetadata(resolvedContext?.trace)
                .copy(context = resolvedContext)
        pushEvent("metric", id, timestamp, resolved.toJsonObject())
    }

    fun action(
        id: String,
        timestamp: String,
        attributes: ActionAttributes,
    ) {
        val resolvedContext = resolvedContext(attributes.context)
        val resolved =
            attributes
                .withResolvedTraceMetadata(resolvedContext?.trace)
                .copy(context = resolvedContext)
        pushEvent("action", id, timestamp, resolved.toJsonObject())
    }

    /** Adds one privacy-bounded breadcrumb to future issue snapshots. */
    fun addBreadcrumb(breadcrumb: IssueBreadcrumb) {
        synchronized(stateLock) {
            requireOpen()
            breadcrumbStore.add(breadcrumb)
        }
    }

    /** Clears all client breadcrumbs and their truncation marker. */
    fun clearBreadcrumbs() {
        synchronized(stateLock) {
            requireOpen()
            breadcrumbStore.clear()
        }
    }

    fun flush(transport: Transport): TransportResponse {
        synchronized(stateLock) {
            if (closed) {
                throw SdkException("shutdown_error", "client is already shut down")
            }
            return flushInternal(transport)
        }
    }

    fun shutdown(transport: Transport): TransportResponse {
        synchronized(stateLock) {
            if (closed) {
                throw SdkException("shutdown_error", "client is already shut down")
            }
            val response = flushInternal(transport)
            closed = true
            return response
        }
    }

    private fun pushEvent(
        type: String,
        id: String,
        timestamp: String,
        attributes: OrderedJsonObject,
    ) {
        Validation.requireNonEmpty("event id", id)
        Validation.requireTimestamp(timestamp)
        synchronized(stateLock) {
            requireOpen()
            events += Event(type, timestamp, id, attributes)
        }
    }

    private fun resolvedContext(eventContext: TelemetryContext?): TelemetryContext? {
        var value = context
        value = TelemetryContext.merge(value, LogBrewTelemetry.currentContext())
        val activeTrace = LogBrewTrace.currentTraceContext()?.toTelemetryTraceContext()
        if (activeTrace != null) {
            value = TelemetryContext.merge(value, TelemetryContext(trace = activeTrace))
        }
        return TelemetryContext.merge(value, eventContext)
    }

    private fun resolvedSpanContext(
        eventContext: TelemetryContext?,
        signalTrace: TelemetryTraceContext?,
    ): TelemetryContext? {
        var value = TelemetryContext.merge(context.withoutTrace(), LogBrewTelemetry.currentContext().withoutTrace())
        value = TelemetryContext.merge(value, eventContext.withoutTrace())
        return TelemetryContext.merge(value, signalTrace?.let { TelemetryContext(trace = it) })
    }

    private fun requireOpen() {
        if (closed) {
            throw SdkException("shutdown_error", "client is already shut down")
        }
    }

    private fun flushInternal(transport: Transport): TransportResponse {
        if (events.isEmpty()) {
            return TransportResponse(204, 0)
        }

        val body = previewJsonLocked()
        val maxAttempts = maxRetries + 1
        for (attempt in 1..maxAttempts) {
            try {
                val response = transport.send(apiKey, body)
                if (response.statusCode == 401) {
                    throw SdkException("unauthenticated", "transport rejected the API key")
                }
                if (response.statusCode in 200..299) {
                    events.clear()
                    return response.copy(attempts = attempt)
                }
                if (response.statusCode >= 500 && attempt < maxAttempts) {
                    continue
                }
                throw SdkException("transport_error", "unexpected transport status ${response.statusCode}")
            } catch (error: TransportException) {
                if (error.retryable && attempt < maxAttempts) {
                    continue
                }
                throw SdkException(error.code, error.message ?: error.code)
            }
        }

        throw SdkException("transport_error", "exhausted retries")
    }

    companion object {
        internal val severityValues = setOf("trace", "debug", "info", "warn", "warning", "error", "fatal", "critical")
        internal val spanStatuses = setOf("ok", "error")
        internal val actionStatuses = setOf("queued", "running", "success", "failure")
        internal val metricKinds = setOf("counter", "gauge", "histogram")
        internal val instantTemporality = setOf("instant")
        internal val deltaCumulativeTemporalities = setOf("delta", "cumulative")

        @JvmOverloads
        fun create(
            apiKey: String,
            sdkName: String,
            sdkVersion: String,
            maxRetries: Int = 2,
            context: TelemetryContext? = null,
            includeAutomaticContext: Boolean = true,
        ): LogBrewClient {
            Validation.requireNonEmpty("api_key", apiKey)
            Validation.requireNonEmpty("sdk_name", sdkName)
            Validation.requireNonEmpty("sdk_version", sdkVersion)
            if (maxRetries < 0) {
                throw SdkException("validation_error", "max_retries must be non-negative")
            }
            val baseContext =
                TelemetryContext.merge(
                    if (includeAutomaticContext) automaticTelemetryContext() else null,
                    context,
                )
            return LogBrewClient(apiKey, sdkName, sdkVersion, maxRetries, baseContext)
        }
    }
}

private fun TelemetryContext?.withoutTrace(): TelemetryContext? {
    if (this == null || trace == null) {
        return this
    }
    if (resource == null && session == null && subject == null && tags == null) {
        return null
    }
    return copy(trace = null).normalized()
}

private fun IssueAttributes.withResolvedTraceMetadata(trace: TelemetryTraceContext?): IssueAttributes =
    copy(metadata = resolvedTraceMetadata(metadata, trace))

private fun LogAttributes.withResolvedTraceMetadata(trace: TelemetryTraceContext?): LogAttributes =
    copy(metadata = resolvedTraceMetadata(metadata, trace))

private fun MetricAttributes.withResolvedTraceMetadata(trace: TelemetryTraceContext?): MetricAttributes =
    copy(metadata = resolvedTraceMetadata(metadata, trace))

private fun ActionAttributes.withResolvedTraceMetadata(trace: TelemetryTraceContext?): ActionAttributes =
    copy(metadata = resolvedTraceMetadata(metadata, trace))

private fun SpanAttributes.telemetryTraceContext(): TelemetryTraceContext? {
    val active = LogBrewTrace.currentTraceContext()
    if (
        active != null &&
        active.traceId.equals(traceId, ignoreCase = true) &&
        active.spanId.equals(spanId, ignoreCase = true) &&
        active.parentSpanId.equals(parentSpanId, ignoreCase = true)
    ) {
        return active.toTelemetryTraceContext()
    }
    return try {
        TelemetryContext(
            trace =
                TelemetryTraceContext(
                    traceId = traceId,
                    spanId = spanId,
                    parentSpanId = parentSpanId,
                ),
        ).normalized().trace
    } catch (_: SdkException) {
        null
    }
}

private fun resolvedTraceMetadata(
    metadata: Map<String, Any?>,
    trace: TelemetryTraceContext?,
): Map<String, Any?> {
    if (trace == null) {
        return metadata
    }
    val traceKeys = setOf("traceId", "spanId", "parentSpanId", "traceFlags", "traceSampled")
    val merged = linkedMapOf<String, Any?>()
    metadata.forEach { (key, value) -> if (key !in traceKeys) merged[key] = value }
    merged["traceId"] = trace.traceId
    trace.spanId?.let { merged["spanId"] = it }
    trace.parentSpanId?.let { merged["parentSpanId"] = it }
    trace.sampled?.let { sampled ->
        merged["traceFlags"] = if (sampled) "01" else "00"
        merged["traceSampled"] = sampled
    }
    return merged.toMap()
}

private data class Event(
    val type: String,
    val timestamp: String,
    val id: String,
    val attributes: OrderedJsonObject,
) {
    fun toJsonObject(): OrderedJsonObject =
        OrderedJsonObject()
            .add("type", type)
            .add("timestamp", timestamp)
            .add("id", id)
            .add("attributes", attributes)
}

File: unity/logbrew-unity/Runtime/LogBrewClient.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace LogBrew.Unity
{
    public sealed class LogBrewClient
    {
        internal static readonly string[] IssueLevels = { "info", "warning", "error", "critical" };
        internal static readonly string[] LogLevels = { "debug", "info", "warning", "error" };
        internal static readonly string[] SpanStatuses = { "ok", "error" };
        internal static readonly string[] ActionStatuses = { "queued", "running", "success", "failure" };
        internal static readonly string[] MetricKinds = { "counter", "gauge", "histogram" };
        internal static readonly string[] InstantTemporality = { "instant" };
        internal static readonly string[] DeltaCumulativeTemporalities = { "delta", "cumulative" };

        private readonly string apiKey;
        private readonly OrderedJsonObject sdk;
        private readonly int maxRetries;
        private readonly TelemetryContext? context;
        private readonly object stateLock = new object();
        private readonly List<Event> events;
        private readonly List<IssueBreadcrumb> breadcrumbs = new List<IssueBreadcrumb>();
        private bool breadcrumbsTruncated;
        private bool closed;

        private LogBrewClient(string apiKey, string sdkName, string sdkVersion, int maxRetries, TelemetryContext? context)
        {
            this.apiKey = apiKey;
            this.maxRetries = maxRetries;
            this.context = context;
            events = new List<Event>();
            sdk = new OrderedJsonObject()
                .Add("name", sdkName)
                .Add("language", "unity")
                .Add("version", sdkVersion);
        }

        public static LogBrewClient Create(
            string apiKey,
            string gameName,
            string sdkVersion,
            int maxRetries = 2,
            TelemetryContext? context = null,
            bool includeAutomaticContext = true)
        {
            Validation.RequireNonEmpty("api_key", apiKey);
            Validation.RequireNonEmpty("game_name", gameName);
            Validation.RequireNonEmpty("sdk_version", sdkVersion);
            if (maxRetries < 0)
            {
                throw new SdkException("validation_error", "max_retries must be non-negative");
            }

            var baseContext = TelemetryContext.Merge(
                includeAutomaticContext ? TelemetryContext.RuntimeDefaults() : null,
                context);
            return new LogBrewClient(apiKey, gameName, sdkVersion, maxRetries, baseContext);
        }

        public int PendingEvents()
        {
            lock (stateLock)
            {
                return events.Count;
            }
        }

        public string PreviewJson()
        {
            lock (stateLock)
            {
                return PreviewJsonLocked();
            }
        }

        public void Release(string id, string timestamp, ReleaseAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent("release", id, timestamp, AttributesWithContext(attributes.ToJsonObject(), resolvedContext));
        }

        public void Environment(string id, string timestamp, EnvironmentAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent("environment", id, timestamp, AttributesWithContext(attributes.ToJsonObject(), resolvedContext));
        }

        public void Issue(string id, string timestamp, IssueAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            List<IssueBreadcrumb> breadcrumbSnapshot;
            bool snapshotTruncated;
            lock (stateLock)
            {
                breadcrumbSnapshot = new List<IssueBreadcrumb>(breadcrumbs);
                snapshotTruncated = breadcrumbsTruncated;
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent(
                "issue",
                id,
                timestamp,
                AttributesWithContextAndTrace(attributes.ToJsonObject(breadcrumbSnapshot.AsReadOnly(), snapshotTruncated), resolvedContext));
        }

        public void Log(string id, string timestamp, LogAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent("log", id, timestamp, AttributesWithContextAndTrace(attributes.ToJsonObject(), resolvedContext));
        }

        public void Span(string id, string timestamp, SpanAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedSpanContext(attributes);
            PushEvent("span", id, timestamp, AttributesWithContext(attributes.ToJsonObject(), resolvedContext));
        }

        public void Metric(string id, string timestamp, MetricAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent("metric", id, timestamp, AttributesWithContextAndTrace(attributes.ToJsonObject(), resolvedContext));
        }

        public void Action(string id, string timestamp, ActionAttributes attributes)
        {
            if (attributes == null)
            {
                throw new ArgumentNullException(nameof(attributes));
            }

            var resolvedContext = ResolvedContext(attributes.Context);
            PushEvent("action", id, timestamp, AttributesWithContextAndTrace(attributes.ToJsonObject(), resolvedContext));
        }

        public void AddBreadcrumb(IssueBreadcrumb breadcrumb)
        {
            if (breadcrumb == null)
            {
                throw new ArgumentNullException(nameof(breadcrumb));
            }

            breadcrumb.ToJsonObject();
            lock (stateLock)
            {
                RequireOpen();
                if (breadcrumbs.Count == IssueDiagnostics.MaxBreadcrumbs)
                {
                    breadcrumbs.RemoveAt(0);
                    breadcrumbsTruncated = true;
                }

                breadcrumbs.Add(breadcrumb);
            }
        }

        public void ClearBreadcrumbs()
        {
            lock (stateLock)
            {
                RequireOpen();
                breadcrumbs.Clear();
                breadcrumbsTruncated = false;
            }
        }

        public TransportResponse Flush(ITransport transport)
        {
            if (transport == null)
            {
                throw new ArgumentNullException(nameof(transport));
            }

            lock (stateLock)
            {
                RequireOpen();
                return FlushInternal(transport);
            }
        }

        public TransportResponse Shutdown(ITransport transport)
        {
            if (transport == null)
            {
                throw new ArgumentNullException(nameof(transport));
            }

            lock (stateLock)
            {
                RequireOpen();
                var response = FlushInternal(transport);
                closed = true;
                return response;
            }
        }

        private void PushEvent(string type, string id, string timestamp, OrderedJsonObject attributes)
        {
            Validation.RequireNonEmpty("event id", id);
            Validation.RequireTimestamp(timestamp);
            lock (stateLock)
            {
                RequireOpen();
                events.Add(new Event(type, timestamp, id, attributes));
            }
        }

        private string PreviewJsonLocked()
        {
            return JsonWriter.Write(new OrderedJsonObject()
                .Add("sdk", sdk)
                .Add("events", events.Select(item => item.ToJsonObject()).ToList()));
        }

        private TelemetryContext? ResolvedContext(TelemetryContext? eventContext)
        {
            var resolved = TelemetryContext.Merge(context, LogBrewTelemetry.CurrentContext);
            var activeTrace = LogBrewTrace.Current;
            if (activeTrace != null)
            {
                resolved = TelemetryContext.WithTrace(resolved, activeTrace);
            }

            return TelemetryContext.Merge(resolved, eventContext);
        }

        private TelemetryContext? ResolvedSpanContext(SpanAttributes attributes)
        {
            var resolved = TelemetryContext.Merge(context?.WithoutTrace(), LogBrewTelemetry.CurrentContext?.WithoutTrace());
            resolved = TelemetryContext.Merge(resolved, attributes.Context?.WithoutTrace());
            try
            {
                var activeTrace = LogBrewTrace.Current;
                var sampled = activeTrace != null
                    && string.Equals(activeTrace.TraceId, attributes.TraceId, StringComparison.OrdinalIgnoreCase)
                    && string.Equals(activeTrace.SpanId, attributes.SpanId, StringComparison.OrdinalIgnoreCase)
                    ? activeTrace.Sampled
                    : (bool?)null;
                var trace = TelemetryContext.Create()
                    .WithTrace(attributes.TraceId, attributes.SpanId, attributes.ParentSpanId, sampled)
                    .Build();
                return TelemetryContext.Merge(resolved, trace);
            }
            catch (SdkException)
            {
                return resolved;
            }
        }

        private static OrderedJsonObject AttributesWithContext(OrderedJsonObject attributes, TelemetryContext? resolvedContext)
        {
            var result = new OrderedJsonObject();
            foreach (var item in attributes.Values)
            {
                if (!string.Equals(item.Key, "context", StringComparison.Ordinal))
                {
                    result.Add(item.Key, item.Value);
                }
            }

            result.AddIfNotNull("context", resolvedContext?.ToJsonObject());
            return result;
        }

        private static OrderedJsonObject AttributesWithContextAndTrace(OrderedJsonObject attributes, TelemetryContext? resolvedContext)
        {
            return AttributesWithContext(LogBrewTrace.AddContextTraceMetadata(attributes, resolvedContext), resolvedContext);
        }

        private void RequireOpen()
        {
            if (closed)
            {
                throw new SdkException("shutdown_error", "client is already shut down");
            }
        }

        private TransportResponse FlushInternal(ITransport transport)
        {
            if (events.Count == 0)
            {
                return new TransportResponse(204, 0);
            }

            var body = PreviewJsonLocked();
            var maxAttempts = maxRetries + 1;
            for (var attempt = 1; attempt <= maxAttempts; attempt++)
            {
                try
                {
                    var response = transport.Send(apiKey, body);
                    if (response.StatusCode == 401)
                    {
                        throw new SdkException("unauthenticated", "transport rejected the API key");
                    }

                    if (response.StatusCode >= 200 && response.StatusCode < 300)
                    {
                        events.Clear();
                        return new TransportResponse(response.StatusCode, attempt);
                    }

                    if (response.StatusCode >= 500 && attempt < maxAttempts)
                    {
                        continue;
                    }

                    throw new SdkException("transport_error", "unexpected transport status " + response.StatusCode.ToString(CultureInfo.InvariantCulture));
                }
                catch (TransportException error)
                {
                    if (error.Retryable && attempt < maxAttempts)
                    {
                        continue;
                    }

                    throw new SdkException(error.Code, error.Message);
                }
            }

            throw new SdkException("transport_error", "exhausted retries");
        }
    }

    internal sealed class Event
    {
        private readonly string type;
        private readonly string timestamp;
        private readonly string id;
        private readonly OrderedJsonObject attributes;

        internal Event(string type, string timestamp, string id, OrderedJsonObject attributes)
        {
            this.type = type;
            this.timestamp = timestamp;
            this.id = id;
            this.attributes = attributes;
        }

        internal OrderedJsonObject ToJsonObject()
        {
            return new OrderedJsonObject()
                .Add("type", type)
                .Add("timestamp", timestamp)
                .Add("id", id)
                .Add("attributes", attributes);
        }
    }
}
